src/pool/srv_util.c

```diff
@@ -277,93 +277,139 @@ ds_pool_transfer_map_buf(struct pool_buf *map_buf, uint32_t map_version,
+	to_remove = d_rank_list_alloc(0);
+	if (to_keep == NULL || to_add == NULL || to_remove == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
 	}
 
-	failed->rl_nr = 0;
-	failed->rl_ranks = NULL;
+	/*
+	 * Fill to_keep and to_remove with all UP and DOWN replicas,
+	 * respectively.
+	 */
+	for (i = 0; i < replicas->rl_nr; i++) {
+		d_rank_list_t  *list;
+		int		j;
 
-	if (nfailed == 0) {
-		alt->rl_nr = 0;
-		alt->rl_ranks = NULL;
-		return 0;
+		for (j = 0; j < nnodes; j++)
+			if (nodes[j].do_comp.co_rank == replicas->rl_ranks[i])
+				break;
+		if (j == nnodes) /* not found (hypothetical) */
+			list = to_remove;
+		else if (map_ranks_include(MAP_RANKS_DOWN, nodes[j].do_comp.co_status)) /* DOWN */
+			list = to_remove;
+		else /* UP */
+			list = to_keep;
+		rc = d_rank_list_append(list, replicas->rl_ranks[i]);
+		if (rc != 0)
+			goto out;
 	}
 
-	/** Make `alt` point to failed subset towards the end **/
-	alt->rl_nr = nfailed;
-	alt->rl_ranks = replicas->rl_ranks + (replicas->rl_nr - nfailed);
-
-	/** Copy failed ranks to make room for replacements **/
-	rc = daos_rank_list_copy(failed, alt);
-	if (rc != 0)
-		return rc;
-
-	/**
-	 * For replacements, search all ranks which are marked as UP
-	 * in the pool map and not present in the list of replicas.
-	 **/
-	for (i = 0, nreplaced = 0; i < nnodes && nreplaced < nfailed; i++) {
-		if (!map_ranks_include(MAP_RANKS_UP,
-				       nodes[i].do_comp.co_status))
-			continue;
-		if (daos_rank_list_find(replicas,
-					nodes[i].do_comp.co_rank, &idx))
-			continue;
-		alt->rl_ranks[nreplaced++] = nodes[i].do_comp.co_rank;
+	/* Compare the UP replicas with the objective. */
+	objective = ds_pool_svc_rf_to_nreplicas(svc_rf);
+	D_ASSERT(to_keep->rl_nr > 0);
+	D_ASSERTF(to_add->rl_nr == 0, "to_add->rl_nr: %u\n", to_add->rl_nr);
+	if (to_keep->rl_nr < objective) {
+		/* More replicas needed. Add new ranks to to_add. */
+		for (i = 0; i < nnodes; i++) {
+			if (!map_ranks_include(MAP_RANKS_UP, nodes[i].do_comp.co_status))
+				continue;
+			/* May be found in replicas but never in to_add. */
+			if (d_rank_list_find(replicas, nodes[i].do_comp.co_rank, NULL /* idx */))
+				continue;
+			rc = d_rank_list_append(to_add, nodes[i].do_comp.co_rank);
+			if (rc != 0)
+				goto out;
+			if (to_add->rl_nr + to_keep->rl_nr == objective)
+				break;
+		}
+		if ((to_add->rl_nr + to_keep->rl_nr) % 2 == 0) {
+			/*
+			 * Not enough engines available, for the objective is
+			 * odd. Do with what we have, but avoid using an even
+			 * number of replicas, which are inefficient.
+			 */
+			if (to_add->rl_nr > 0) {
+				to_add->rl_nr--;
+			} else {
+				d_rank_t victim = to_keep->rl_ranks[to_keep->rl_nr - 1];
+
+				rc = d_rank_list_append(to_remove, victim);
+				if (rc != 0)
+					goto out;
+				to_keep->rl_nr--;
+			}
+		}
+	} else if (to_keep->rl_nr > objective) {
+		/*
+		 * Too many replicas---perhaps, since we may be using
+		 * svc_rf_default. Hence, we don't reduce to the objective, but
```

Comment:
Sorry for the obscurity. How about the following?

		/*
		 * Too many replicas. Since the objective may be based on
		 * svc_rf_default (see [Temporary] in the function
		 * documentation), do not reduce the number of replicas to the
		 * objective. If the number of replicas is even, however,
		 * remove a replica to make the number odd, because an even
		 * number is inefficient.
		 */

Reply:
No problem at all. This looks good.